### PR TITLE
CLN support

### LINKED ISSUE #196: Add core-lightning as LN node provider.
**Is your feature request related to a problem? Please describe.**
We are currently relying on [LND](https://github.com/lightningnetwork/lnd) for RoboSats backend node. It does work pretty well. There is only a few things that might be critical: channel.db is hard to manage and performance/reliability could still be better (I guess?). 

It would be great to implement [core-lightning](https://github.com/ElementsProject/lightning) as backend, not as a replacement, but for exploration and extra redundancy. The higher performance and use of Postgres as database would be much appreciated. While it is not clear yet whether there will be some shenanigan or unexpected issue, the redundancy of having two node providers will always be of great value.

If core-lightning is available as node provider, future effort to improve RoboSats availability could be put into creating a "LN load balancer" that will smartly direct invoice creation / payments to a pool of nodes of different providers.

**Describe the solution you'd like**

We need a new [`api/lightning/node.py`](https://github.com/Reckless-Satoshi/robosats/blob/main/api/lightning/node.py) (LNnode provider) that can perform all of the same functions using core-lightning as backend instead of LND. 

- [ ] **Prerequisites not met** By @louneskmt we need a hodl invoice plugin in Core-Lightning. Out of scope for RoboSats.
- [ ] The first step would be to add a [core-lightning container](https://hub.docker.com/r/elementsproject/lightningd/tags) running testnet3 to our dev [docker-compose](https://github.com/Reckless-Satoshi/robosats/blob/main/docker-compose.yml) orchestration. 
- [ ] It might be useful to implement unit tests `test.node.py` to learn about the current set up. The unit tests will be very helpful in the future, as they should receive exactly the same behavior regardless of whether the call was attended by LND or core-lightning  in any future release of these.
- [ ] The bigger chunk of the work: write all core-lightning calls for every function.

I am unsure how big this task really is. But it certainly requires some very specific knowledge on interacting with Lightning node APIs.

**Additional context**
This task is ⚡**rewarded with 2M Sats**⚡ . Drop a comment if you want to be assigned.

### PR DESCRIPTION
**Pull request by daywalker90**
## What does this PR do?
add node_cln.py and proto generation that works together with my new cln-grpc-hodl plugin

This PR introduces an initial attempt at support for cln. Fixes #196.
This pr does NOT include stuff related to docker, tests :) or new and missing config variables.
A few things are unclear (e.g. the TODO's) on how to handle them because cln works a bit different in some places.
Also this is completely untested (me docker noob) but i wanted to show my progress and get some feedback.

See https://community.corelightning.org/c/developers/hold-invoice-plugin#comment_wrapper_17947441 on how to build and install the cln-grpc-hodl plugin or download the binary for amd64_linux here: [cln-grpc-hodl_amd64_linux.zip](https://github.com/daywalker90/lightning/releases/download/cln-grpc-hodl-0.1.1/cln-grpc-hodl_amd64_linux.zip)

### CODE REVIEW

**Comment by daywalker90** [comment]
you need atleast protoc 3.12 i think but also <4 iirc

**Comment by Reckless-Satoshi** [comment]
We got the new service for CLN w/ Hodlvoice in branch `add-cln-hodl-backend` . Tested and working perfectly (the service itself, did not test the plugin). Check #418 for more details. 

Can you rebase this branch over that one? That way we have the pre-requisites ready to start testing the new `node_cln.py` functions, the development robo container can now do requests to the grpc-hodl plugin.

**Comment by daywalker90** [comment]
I found 2 bugs in the plugin today i have to fix first.

File: api/lightning/node_cln.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+        local_balance_sat = 0
+        remote_balance_sat = 0
+        for peer in response.peers:
+            for channel in peer.channels:
+                if channel.state == 2: # CHANNELD_NORMAL
+                    local_balance_sat += channel.to_us_msat.msat // 1_000
+                    remote_balance_sat += (channel.total_msat.msat-channel.to_us_msat.msat) // 1_000
+        
+        # TODO no idea what is meant by unsettled/pending balance here exactly
```
Comment (Reckless-Satoshi) [inline api/lightning/node_cln.py:132]: "Unsettled balance" is the total amount on incoming HTLCs on hold. "Pending balance" is the total "in_flight" amount (outgoing payments in transition).

File: api/lightning/node_cln.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        # The preimage is a random hash of 256 bits entropy
+        preimage = hashlib.sha256(secrets.token_bytes(nbytes=32)).digest()
+
+        # Its hash is used to generate the hold invoice
+        r_hash = hashlib.sha256(preimage).digest()
+
+        request = noderpc.InvoiceRequest(
+            description=description,
+            amount_msat=num_satoshis * 1_000,
+            label="TODO", # TODO needs to be a unique string
```
Comment (Reckless-Satoshi) [inline api/lightning/node_cln.py:207]: A "label" in CLN is some sort of description only for internal use? We could add a couple more parameters to `gen_hodl_invoice()` so this function is informed on `order_id` , `receiver_robot`, `time` and other unique identifiers to build a string for this label that is useful for bookeeping too.

**Review by Reckless-Satoshi — COMMENTED** [review]
I'll spend some more time in the next couple reviewing this class.

File: api/lightning/node_cln.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        if response.state == 3:  # ACCEPTED (LOCKED)
+            lnpayment.expiry_height = response.htlc_cltv
+            lnpayment.status = LNPayment.Status.LOCKED
+            lnpayment.save()
+            return True
+
+    @classmethod
+    def resetmc(cls):
+        # don't think an equivalent exists for cln, maybe deleting gossip_store file?
+        return False
```
Comment (Reckless-Satoshi) [inline api/lightning/node_cln.py:256]: In practice this function is not used. So it is safe to leave it as a blank placeholder.

File: api/lightning/node_cln.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+            return payout
+
+        payout["valid"] = True
+        payout["description"] = payreq_decoded.description
+        payout["payment_hash"] = payreq_decoded.payment_hash
+
+        return payout
+
+    @classmethod
+    def pay_invoice(cls, lnpayment):
```
Comment (Reckless-Satoshi) [inline api/lightning/node_cln.py:350]: We have a pseudo-duplicate of this function outside the LNNode class. This one here used to pay out rewards. The other one, more sophisticated, is in `api/tasks` named `follow_send_payment(hash)`. 

While `api/lightning/node.py` pay_invoice() simply attempts to pay and expects a succeed/failed, the one in `tasks.py` evaluates every update during the routing (it uses an async celery worker thread). By using many workers and checking over all payments continuously the payout process is more robust to routing failures always keeping the user in the loop. There are also many shenanigans (e.g., payment is hodl by the recipient) that `follow_send_payment()` the celery worker task is safer for.

We will probably need to refactor and move the core gRPC functionality of `follow_send_payment()` into the `LNNode` class, so the celery task can use both CLN or LND.

File: generate_grpc.sh [hunk 5]
```diff
@@ -10,6 +10,11 @@ python3 -m grpc_tools.protoc --proto_path=googleapis:. --python_out=. --grpc_pyt
 curl -o router.proto -s https://raw.githubusercontent.com/lightningnetwork/lnd/master/lnrpc/routerrpc/router.proto
 python3 -m grpc_tools.protoc --proto_path=googleapis:. --python_out=. --grpc_python_out=. router.proto
 
+# generate cln grpc definitions
+curl -o node.proto -s https://raw.githubusercontent.com/daywalker90/lightning/hodlvoice/cln-grpc/proto/node.proto
+curl -o primitives.proto -s https://raw.githubusercontent.com/daywalker90/lightning/hodlvoice/cln-grpc/proto/primitives.proto
+python3 -m grpc_tools.protoc --proto_path=. --python_out=. --grpc_python_out=. node.proto primitives.proto
```
Comment (Reckless-Satoshi) [inline generate_grpc.sh:16]: Nice!

**Review by Reckless-Satoshi — COMMENTED** [review]
I'm impressed by everything that went into creating this new CLN enabled LNNode class. Congratulations on the work! Sure, it looks like we are going to need to iterate a lot during testing, but it's exciting to add CLN to the stack :)

I will write here some things that will still be missing to have a functionally CLN-enabled RoboSats coordinator so we do not skip over them:
- [ ] In `/api/utils` we have another function that should belong in LNNode class, that is, `get_lnd_version()`. We could add there another one, `get_cln_version()` that makes use of a new `LNNode` class function to query the version.
- [ ] Add a new response parameter `/api/info` endpoint with `cln_version`.
- [ ] Add a switch for LN node backend and a new environmental variable `LN_NODE = LND | CLN`
- [ ] Optionally: add a prototype dummy wrapper (or "load balancer") class that allow us to have both LND and CLN running at the same time. Does not need to do anything clever, just a wrapper so we can implement clever things down the line (...down the line, ideally, we introduce a new Model `LN_Node`, each node in the stack will have its own DB object with fields: node_vendor CLN | LND, version, last balance_log, etc. Each `LNPayment` will reference the `LN_Node` that created it, therefore the robosats API can manage invoices incoming/outgoing through multiple nodes. This is very important for high availability and uptime.)

For the client:
- [ ] Add new a item to the Coordinator stats for needs dialog to display CLN node version.

File: api/lightning/node_cln.py (hunk 2 above)
Comment (daywalker90) [reply api/lightning/node_cln.py]: done

File: api/lightning/node_cln.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        cls, num_satoshis, description, invoice_expiry, cltv_expiry_blocks
+    ):
+        """Generates hold invoice"""
+
+        hold_payment = {}
+        # The preimage is a random hash of 256 bits entropy
+        preimage = hashlib.sha256(secrets.token_bytes(nbytes=32)).digest()
+
+        # Its hash is used to generate the hold invoice
+        r_hash = hashlib.sha256(preimage).digest()
```
[the remark this answers is not in the document]
Comment (daywalker90) [reply api/lightning/node_cln.py]: done

**Comment by daywalker90** [comment]
Added draft for follow_send_payment. Again, untested.

**Comment by Reckless-Satoshi** [comment]
Hey @daywalker90 the base branch `add-cln-hodl-backend` has some new commits. 

All of the LND gRPC related calls are now in the LNNode class (two new functions added `follow_send_payment()` and `lookup_invoice_status()`. The LNNode class for LND is now called `LNDNode` and for the one for CLN is `CLNNode`.  A new environmental variable "LNNODE=LND | CLN" is used to import one or the other as `LNNode`.

The LND version is now fetched from RPC, a new proto was added, so there will be some conflicts on `generate_grpc.sh` once you rebase your work on top of that branch. You'll also have to move your LNNode class to `api/lightning/cln.py`.  In addition the frontend should also now show the CLN version if everything works well.

**Comment by daywalker90** [comment]
did all that and added the 2 functions

**Comment by daywalker90** [comment]
clns gen_hold_invoice has more parameters for its label than lnds gen_hold_invoice right now

**Comment by Reckless-Satoshi** [comment]
Hey @daywalker90 the work in this branch is now merged to `add-cln-hodl-backend` https://github.com/Reckless-Satoshi/robosats/pull/433

We can keep building / testing / debugging with small PRs to `add-cln-hodl-backend` until the thing is ready to be merged into `main`.

Closing this one.

**Pull request closed without merging**